sahi/utils/cv.py

```diff
@@ -210,6 +210,78 @@ def apply_color_mask(image: np.ndarray, color: tuple):
     return colored_mask
 
 
+def get_video_reader(
+    source: str, save_dir: str, export_visual: bool = False, view_image: bool = False, fast_forwarding: bool = False
+):
+    # get video name with extension
+    image_base = os.path.basename(source)
+    # get video from video path
+    video_capture = cv2.VideoCapture(source)
+    if export_visual:
+        # get video properties and create VideoWriter object
+        fps = video_capture.get(cv2.CAP_PROP_FPS)
+        w = int(video_capture.get(cv2.CAP_PROP_FRAME_WIDTH))
+        h = int(video_capture.get(cv2.CAP_PROP_FRAME_HEIGHT))
+        size = (w, h)
+        fourcc = cv2.VideoWriter_fourcc(*"mp4v")
+        out = cv2.VideoWriter(os.path.join(save_dir, image_base), fourcc, fps, size)
+
+        if view_image:
+            cv2.imshow("Prediction of {}".format(str(image_base)), cv2.WINDOW_AUTOSIZE)
+
+    if fast_forwarding:
+
+        def read_video_frame(video_capture):
+            while video_capture.isOpened:
+                k = cv2.waitKey(30)
+
+                frame_num = video_capture.get(cv2.CAP_PROP_POS_FRAMES)
+
+                if k == 27:
+                    print(
+                        "\n===========================Closing==========================="
+                    )  # Exit the prediction, Key = Esc
+                    exit()
+                if k == 100:
+                    frame_num += 100  # Skip 100 frames, Key = d
+                if k == 97:
+                    frame_num -= 100  # Prev 100 frames, Key = a
+                if k == 103:
+                    frame_num += 10  # Skip 10 frames, Key = g
+                if k == 102:
+                    frame_num -= 10  # Prev 10 frames, Key = f
+                video_capture.set(cv2.CAP_PROP_POS_FRAMES, frame_num)
+
+                ret, frame = video_capture.read()
+                if not ret:
+                    print("\n===========================Video Ended===========================")
+                    break
+                yield Image.fromarray(frame)
+
+    else:
+
+        def read_video_frame(video_capture):
+            while video_capture.isOpened:
+                k = cv2.waitKey(30)
+
+                if k == 27:
+                    print(
+                        "\n===========================Closing==========================="
+                    )  # Exit the prediction, Key = Esc
+                    exit()
+
+                ret, frame = video_capture.read()
+                if not ret:
+                    print("\n===========================Video Ended===========================")
+                    break
+                yield Image.fromarray(frame)
```

Comment:
Thanks for the refactor. My revision here is again, to refactor this inner functions/iterators. I suggest you to create the function `read_video_frame` (independent of this scope) another util function, which takes `video_capture` and `fast_forwarding` as parameters. For example it'd be sth like this,

```python
def read_video_frame(video_capture, fast_forwarding):
    if fast_forwarding:
        ...
        yield ...
    else:
        ...
        yield ...
```

Reply:
I correct it.

Reply:
Here, I was thinking that, why would not we make the fast-forwarding the default and only behavior, does it impose any harm to the users ? wdyt about making it the only way @madenburak @fcakyon ? Fast-forwarding mode already covers the usage of not using the FF mode, or is there any ?

Reply:
You're right. I thinked without FF program is Faster than with FF. But When I reviewed, this situation is unreasonable. FF must be default choice. Also, If num of slice is increase, view image is slowing. For this, I thinking, we can process some frames(users choice) instead of all frames. For examle, one of every 3(user’s choice) frames. If user don’t want to this, can choice 1.